### PR TITLE
Dedicated threadpool for system index writes

### PR DESCRIPTION
**Pull request by jaymode**
This commit adds a dedicated threadpool for system index write
operations. The dedicated resources for system index writes serves as
a means to ensure that user activity does not block important system
operations from occurring such as the management of users and roles.

### CODE REVIEW

**Comment by tbrooks8** [comment]
You might want to force execution on `IndexingPressure` otherwise too much user memory usage will block these system operations.

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java [hunk 1]
```diff
@@ -61,21 +62,25 @@
         > extends TransportReplicationAction<Request, ReplicaRequest, Response> {
 
     private final IndexingPressure indexingPressure;
-    private final String executor;
+    private final Function<IndexShard, String> executor;
```
Comment (gwbrown) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:65]: Nitpick: the name `executor` isn't really right now that it's actually a function to get an executor.

File: server/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionIngestTests.java [hunk 2]
```diff
@@ -274,7 +276,7 @@ public void testIngestLocal() throws Exception {
         assertFalse(responseCalled.get());
         assertFalse(failureCalled.get());
         verify(ingestService).executeBulkRequest(eq(bulkRequest.numberOfActions()), bulkDocsItr.capture(),
-            failureHandler.capture(), completionHandler.capture(), any());
+            failureHandler.capture(), completionHandler.capture(), any(), eq(Names.WRITE));
```
Comment (gwbrown) [inline server/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionIngestTests.java:279]: There are a number of instances where we now verify that we're using the `WRITE` threadpool here. Could we add a test here to check the logic that determines when to use `SYSTEM_WRITE`?

**Review by AthenaEryma — APPROVED** [review]
Left a few minor comments, LGTM otherwise (no need for another round of review once the comments are addressed). You may want a second opinion from someone more familiar with this code though - write threadpools are not my area of expertise.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 3]
```diff
@@ -336,6 +345,18 @@ static void prohibitCustomRoutingOnDataStream(DocWriteRequest<?> writeRequest, M
         }
     }
 
+    boolean isOnlySystem(BulkRequest request, SortedMap<String, IndexAbstraction> indicesLookup, SystemIndices systemIndices) {
+        final boolean onlySystem = request.getIndices().stream().allMatch(indexName -> {
+            final IndexAbstraction abstraction = indicesLookup.get(indexName);
+            if (abstraction != null) {
+                return abstraction.isSystem();
+            } else {
+                return systemIndices.isSystemIndex(indexName);
```
[the remark this answers is not in the document]
Comment (jaymode) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:354]: At the point when we need to call this, we cannot guarantee that all system indices have been created :(

**Review by Tim-Brooks — APPROVED** [review]
LGTM - I only looked at the indexing pressure stuff in bulk action and shard bulk action. If you wanted a full review from me, let me know.

**Pull request merged**